swap/protocol.go

```diff
@@ -0,0 +1,112 @@
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+// ErrEmptyAddressInSignature is used when the empty address is used for the chequebook in the handshake
+var ErrEmptyAddressInSignature = errors.New("empty address in handshake")
+
+var Spec = &protocols.Spec{
+	Name:       "swap",
+	Version:    1,
+	MaxMsgSize: 10 * 1024 * 1024,
+	Messages: []interface{}{
+		SwapHandshakeMsg{},
+		EmitChequeMsg{},
+		ErrorMsg{},
+		ConfirmMsg{},
+	},
+}
+
+func (s *Swap) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{
+		{
+			Name:    Spec.Name,
+			Version: Spec.Version,
+			Length:  Spec.Length(),
+			Run:     s.run,
+		},
+	}
+}
+
+func (s *Swap) APIs() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: "swap",
+			Version:   "1.0",
+			Service:   s.api,
+			Public:    false,
+		},
+	}
+}
+
+func (s *Swap) Start(server *p2p.Server) error {
+	log.Info("Swap service started")
+	return nil
+}
+
+func (s *Swap) Stop() error {
+	return nil
+}
+
+// verifyHandshake verifies the chequebook address transmitted in the swap handshake
+func (s *Swap) verifyHandshake(msg interface{}) error {
+	handshake, ok := msg.(*SwapHandshakeMsg)
+	if !ok || (handshake.ContractAddress == common.Address{}) {
+		return ErrEmptyAddressInSignature
+	}
+
+	return s.verifyContract(context.TODO(), handshake.ContractAddress)
+}
+
+func (s *Swap) run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
+	protoPeer := protocols.NewPeer(p, rw, Spec)
+
+	answer, err := protoPeer.Handshake(context.TODO(), &SwapHandshakeMsg{
+		ContractAddress: s.owner.Contract,
+	}, s.verifyHandshake)
+	if err != nil {
+		return err
+	}
+
+	beneficiary, err := s.getContractOwner(context.TODO(), answer.(*SwapHandshakeMsg).ContractAddress)
+	if err != nil {
+		return err
+	}
+
+	swapPeer := NewPeer(protoPeer, s, s.backend, beneficiary, answer.(*SwapHandshakeMsg).ContractAddress)
+
+	s.lock.Lock()
+	s.peers[p.ID()] = swapPeer
+	s.lock.Unlock()
+
+	s.logBalance(protoPeer)
+
+	return swapPeer.Run(swapPeer.handleMsg)
+}
+
+type PublicAPI struct {
```

Comment:
@zelig why did you decide to use `Swap` as the main Swap service API provider?
I had `Service` because of the potential to shield some methods from RPC - Now as I understand, even `Add` can be called via RPC, which I think shouldn't be?

Also, what is this `PublicAPI` for ?

Reply:
Well this is code which came out of the Balaton week so I have no idea what this empty `PublicAPI` struct is what it should be used for.

Disregard the first part, it is now implemented as `SwapInfo`

Reply:
https://github.com/ethersphere/swarm/issues/1591 will take care of this